libnetwork: cleanup config package, remove old integration tests

thaJeztah opened the pull request:
see individual commits for details

thaJeztah committed:
libnetwork/config: remove "Experimental" and "Debug" options

These were no longer used.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/config: inline ProcessOptions

This method was only used in a single place; inlining it makes it
easier to see what's done.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/config: remove ParseConfig()

Libnetwork configuration files were only used as part of integration tests using
the dnet utility, which was removed in 7266a956a88228cd5aed26894b296ad4ffc6e2f6

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: remove old integration tests

This was used for testing purposes when libnetwork was in a separate repo, using
the dnet utility, which was removed in 7266a956a88228cd5aed26894b296ad4ffc6e2f6.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/config: merge DaemonCfg into Config

It was unclear what the distinction was between these configuration
structs, so merging them to simplify.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/config: inline LoadDefaultScopes()

This method was an exported method, but only used as part of ParseConfigOptions,
so inlining it.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: fix empty-lines (revive)

    libnetwork/etchosts/etchosts_test.go:167:54: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/osl/route_linux.go:185:74: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/osl/sandbox_linux_test.go:323:36: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/bitseq/sequence.go:412:48: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/datastore/datastore_test.go:67:46: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/datastore/mock_store.go:34:60: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/iptables/firewalld.go:202:44: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/iptables/firewalld_test.go:76:36: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/iptables/iptables.go:256:67: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/iptables/iptables.go:303:128: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/networkdb/cluster.go:183:72: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/ipams/null/null_test.go:44:38: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/drivers/macvlan/macvlan_store.go:45:52: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/ipam/allocator_test.go:1058:39: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/drivers/bridge/port_mapping.go:88:111: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/drivers/bridge/link.go:26:90: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/drivers/bridge/setup_ipv6_test.go:17:34: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/drivers/bridge/setup_ip_tables.go:392:4: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/drivers/bridge/bridge.go:804:50: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/drivers/overlay/ov_serf.go:183:29: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/drivers/overlay/ov_utils.go:81:64: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/drivers/overlay/peerdb.go:172:67: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/drivers/overlay/peerdb.go:209:67: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/drivers/overlay/peerdb.go:344:89: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/drivers/overlay/peerdb.go:436:63: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/drivers/overlay/overlay.go:183:36: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/drivers/overlay/encryption.go:69:28: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/drivers/overlay/ov_network.go:563:81: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/default_gateway.go:32:43: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/errors_test.go:9:40: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/service_common.go:184:64: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/endpoint.go:161:55: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/store.go:320:33: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/store_linux_test.go:11:38: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/sandbox.go:571:36: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/service_common.go:317:246: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/endpoint.go:550:17: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/sandbox_dns_unix.go:213:106: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/controller.go:676:85: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/agent.go:876:60: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/resolver.go:324:69: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/network.go:1153:92: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/network.go:1955:67: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/network.go:2235:9: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/libnetwork_internal_test.go:336:26: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/resolver_test.go:76:35: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/libnetwork_test.go:303:38: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/libnetwork_test.go:985:46: empty-lines: extra empty line at the end of a block (revive)
    libnetwork/ipam/allocator_test.go:1263:37: empty-lines: extra empty line at the start of a block (revive)
    libnetwork/errors_test.go:9:40: empty-lines: extra empty line at the end of a block (revive)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

corhere commented on libnetwork/config/config.go:
```diff
@@ -34,38 +34,40 @@ type Config struct {
 		Scopes:    make(map[string]*datastore.ScopeCfg),
 	}
+
+	opts = append(opts, OptionDefaultScopes(cfg.DataDir))
```
The value of `cfg.DataDir` is captured before the options are applied, so will always be the zero value. A user-configured value using `OptionDataDir()` would not be respected.

thaJeztah replied:
🤦 you are so right; good catch. I started with inlining the code, then thought it would be nice to keep the GoDoc / description, so I went back and changed it to a `Option`, but completely overlooked this.

As this one is only used internally, I should probably also un-export it, and remove the closure.

thaJeztah replied:
@corhere updated; PTAL

corhere commented on libnetwork/config/config.go:
```diff
@@ -34,38 +34,38 @@ type Config struct {
 		Scopes:    make(map[string]*datastore.ScopeCfg),
 	}
+
+	opts = append(opts, optionDefaultScopes)
```
Nit: refactoring `func (*Config).LoadDefaultScopes` into an `Option` is too clever for my taste. `optionDefaultScopes` is a special case compared to all the other `Option`s: it is the only one which consumes a `Config` field and is therefore sensitive to placement in the `opts` slice, it is unconditionally appended to user-supplied options, and (as far as I can tell) it would not make sense to be passed in by the user if it was exported. And indirecting what is essentially an unconditional function call through the `opts` slice harms readability by obscuring control flow. What advantages of this change am I overlooking?

thaJeztah replied:
@corhere I could go back to my initial approach to inline it; I thought it was slightly cleaner to just append it as an operation to the list of options to re-use the existing logic, but agree that it's unlikely useful to be passed in as an option.

thaJeztah replied:
Updated; I inlined it in `ParseConfigOptions()`

corhere reviewed (approved):
LGTM, aside from my "readability" nitpick for commit 26747f98d8eec501db1a54cccda49e34b346201c.

thaJeztah commented on libnetwork/config/config.go:
```diff
@@ -41,48 +29,31 @@ type DaemonCfg struct {
 // ParseConfigOptions parses the configuration options and returns
 // a reference to the corresponding Config structure
-func ParseConfigOptions(cfgOptions ...Option) *Config {
+func ParseConfigOptions(opts ...Option) *Config {
```
Actually considering if we should rename this to `New()` - effectively it's a constructor for Config; WDYT?

corhere replied:
I think that's a great idea! 👍

thaJeztah replied:
Done! 🎉

thaJeztah committed:
libnetwork/config: rename ParseConfigOptions() to New()

This function effectively is a constructor, so rename it to better describe
it's functionality.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
